agent/engine/dockerstate/docker_task_engine_state.go

```diff
@@ -148,17 +151,28 @@ func (state *DockerTaskEngineState) AllTasks() []*apitask.Task {
 	state.lock.RLock()
 	defer state.lock.RUnlock()
 
-	return state.allTasksUnsafe()
+	return state.allTasksUnsafe(false)
 }
 
-func (state *DockerTaskEngineState) allTasksUnsafe() []*apitask.Task {
+func (state *DockerTaskEngineState) allTasksUnsafe(excludeInternal bool) []*apitask.Task {
```

Comment:
I wouldn't change the prototype here. I would simply create a new function that this calls with false and the new AllExternalTasks calls with true. like getFillteredTasksUnsafe(excludeInternal bool)

Reply:
`allTasksUnsafe` is an internal function, used by the exported `AllTasks`. I added a new exported method `AllExternalTasks`, but want to reuse the `allTasksUnsafe` to iterate through all tasks. So it looks like 
```
AllTasks() 
  -> allTaskUnsafe(false)

AllExternalTasks() 
  -> allTaskUnsafe(true)
```
Is this what you are suggesting?

Reply:
sort of, I see a lot of diffs that are not needed and don't make sense, basically any line that just adds a "false"
```go
AllTasks() 
  -> allTaskUnsafe()
      -> getFilteredTasksUnsafe(false)

AllExternalTasks() 
  -> allExternalTasksUnsafe()
      -> getFilteredTasksUnsafe(false)
```

So the non-Unsafe versions are external yes, but internally the Usafe version is used a lot. The above is a more structured breakdown of responsibilities. External "safe" versions exist, if you already have the lock then use the "unsafe" versions which are exactly equivalent to the "safe" version from an API perspective, but the unsafe versions are 99% the same so they share a common impl in getFilteredTasksUnsafe.

Reply:
Thanks for clarifying. Makes sense to me!